edx_prefectutils/amplitude.py

```diff
@@ -0,0 +1,347 @@
+        # following dimension:
+        #
+        #   coalesce(user_id, device_id)
+        #
+        # Then, sort events by segment size and sort by timestamp within each segment.
+        segmented_by_device_user_query = (
+            f"""
+              select coalesce(user_id, device_id) as user_device,
+                     count(*) over (partition by user_device) as user_device_event_count,
+                     *
+                from events_to_load
+               where amplitude_event_id >= {processing_segment_start}
+                 and amplitude_event_id < {processing_segment_end}
+            order by user_device_event_count asc,
+                     user_device asc,
+                     timestamp asc
+            """
+        )
+        segmented_by_device_user = cursor.execute(segmented_by_device_user_query)
+
+        # Make a generator which can fetch events and slice them for worker threads.
+        segmented_by_device_user_and_batched = _fetch_and_consolidate_and_batch(cursor)
+
+        # Only allow a fixed number (64, for example) of simultaneous network calls.
+        available_senders = threading.Semaphore(64)
+
+        @backoff.on_exception(
+            backoff.expo,
+            (
+                requests.exceptions.HTTPError,  # i.e. the status is 4xx or 5xx.
+                requests.exceptions.ConnectionError,  # This includes connection timeout.
+                requests.exceptions.Timeout,  # This includes read timeout.
+                # In other data pipelines, retrying on read timeout could be dangerous because the events from the
+                # failed request my have been successfully ingested, so retrying can result in duplicate events.
+                # However, we do not mind retrying on read timeout with Amplitude loading because Amplitude will de-dupe
+                # the events anyway.
+            ),
+        )
+        def send_batch(amplitude_events_batch):
+            """
+            Actually make a Batch API request to send a batch of events.
+
+            This is meant to be executed by a worker thread only.  Since we only want to allow a finite number of worker
+            threads to be running API calls simultaneously, the API call is surrounded by logic to acquire/release a
+            semaphore lock.  It's important that this semaphore logic exists *inside* the retry loop because we don't want
+            to block any other workers while this one is in the middle of backing off.
+
+            Args:
+                amplitude_events_batch (iterable of events): 1000 or fewer events formatted for consumption by Amplitude.
+            """
+            available_senders.acquire()
+            r = requests.post(
+                "https://api2.amplitude.com/2/httpapi",
+                json={
+                    "api_key": amplitude_api_key,
+                    "events": amplitude_events_batch,
```

Comment:
nit: I think we also want to set min_user_id_length = 1 or something like that, as an option.